### PR TITLE
GT-1297 GitHub actions

### PR DESCRIPTION
**Pull request by levieggertcru**

### CODE REVIEW

**Comment by levieggertcru** [comment]
@gyasistory @frett let me know what you think.  Everything is in the build.yml and there's just one job.  That seemed to be the simplest without repeating steps and everything can run on macos-11 which chooses Xcode 13.

**Comment by frett** [comment]
> @gyasistory @frett let me know what you think. Everything is in the build.yml and there's just one job. That seemed to be the simplest without repeating steps and everything can run on macos-11 which chooses Xcode 13.

Repeating steps between multiple jobs is something that is annoying in GHA. The only solution I've seen is to define a custom step that contains the shared logic. But that seems like a lot of overhead.

On other repos we have duplicated steps when it makes sense to separate a job into multiple jobs

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -28,33 +22,35 @@ jobs:
       - name: Update Fastlane
         run: bundle update fastlane
```
Comment (frett) [inline .github/workflows/build.yml:23]: you shouldn't need to update fastlane here, this is handled automatically by dependabot or renovate

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -28,33 +22,35 @@ jobs:
...
         run: bundle exec fastlane cru_shared_lane_download_and_commit_latest_one_sky_localizations
 
-      - name: Decode AppleAppStoreApi.json.enc For Fastlane App Store Connect Access 
-        if: github.event_name == 'workflow_dispatch'
+      - name: Decrypt AppleAppStoreApi.json.enc For Fastlane App Store Connect Access 
+        if: github.event_name == 'push' && github.ref == 'refs/heads/master'
         env: 
           KEY: ${{ secrets.ENCRYPTED_CF3A694B4ADE_KEY }}
           IV: ${{ secrets.ENCRYPTED_CF3A694B4ADE_IV }}
         run: openssl aes-256-cbc -K ${KEY} -iv ${IV} -in fastlane/AppleAppStoreApi.json.enc -out fastlane/AppleAppStoreApi.json -d
```
Comment (frett) [inline .github/workflows/build.yml:45]: If the API key is an actual json file, we can just store the contents of that file as a secret instead of having to use openssl to decode the file

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -28,33 +22,35 @@ jobs:
...
 
       - name: Increment Xcode Project Build Number
-        if: github.event_name == 'workflow_dispatch'
+        if: github.event_name == 'push' && github.ref == 'refs/heads/master'
         run: bundle exec fastlane cru_shared_lane_increment_xcode_project_build_number
```
Comment (frett) [inline .github/workflows/build.yml:49]: why did you create new fastlane lanes instead of using the ones we had?

**Review by frett — CHANGES_REQUESTED** [review]
several questions, the big one being, why did you create new fastlane lanes instead of using what we already had defined?

**Comment by frett** [comment]
here is GHA for MissionHub using our already existing fastlane lanes: https://github.com/CruGlobal/missionhub-react-native/blob/develop/.github/workflows/ci-cd.yml#L226-L242

File: .github/workflows/build.yml (hunk 3 above)
Comment (levieggertcru) [reply .github/workflows/build.yml:49]: Hey @frett there are a number of reasons.

Comment (levieggertcru) [reply .github/workflows/build.yml:49]: @frett  I'll gather some notes and include them here.

Comment (levieggertcru) [reply .github/workflows/build.yml:49]: Hey @frett the biggest reason was how code signing and match was done, the cru fastfile is including MPDX's Cal Directory Extension in code signing and match which isn't necessary and would rather opt for a more dynamic approach to this.  That way if we have iOS apps that include other app extensions we don't bloat the cru fastfile.  Also with match you don't need to create a new match for every app extension.  Match supports multiple bundle ids.

Comment (levieggertcru) [reply .github/workflows/build.yml:49]: @frett also some of the naming, for example, in localizations download they're using ```CRU_SCHEME``` which actually has nothing to do with the localizations download, that is actually a root directory name.

Comment (levieggertcru) [reply .github/workflows/build.yml:49]: @frett I do want to look into including a firebase distribution in the cru fastfile as well which would require different code signing than what is used for TestFlight, which is another reason I wanted to break up some of the lanes.

Comment (frett) [reply .github/workflows/build.yml:49]: I definitely agree about not including the MPDX call directory stuff in the shared fastlane file, I hadn't realized a contractor had done that several years ago. I would have definitely pushed more a more configurable option, or at least putting some call directory specific lanes in the MPDX Fastfile.

File: .github/workflows/build.yml (hunk 2 above)
Comment (levieggertcru) [reply .github/workflows/build.yml:45]: Hey @frett I like this idea better, I will look into this.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -28,33 +22,35 @@ jobs:
       - name: Update Fastlane
         run: bundle update fastlane
 
+        # Required for KotlinMultiplatform
+      - uses: actions/setup-java@v1
+        with:
+          java-version: '11'
+
       - name: Pod Install
         run: bundle exec fastlane cru_shared_lane_pod_install
```
Comment (frett) [inline .github/workflows/build.yml:31]: any reason you are preferring to install the pods via fastlane instead of just running `pod install` here?

File: .github/workflows/build.yml (hunk 3 above)
Comment (frett) [reply .github/workflows/build.yml:49]: ok, I'm going to process the new lanes a bit, it would be nice if there was a PR with the new lanes so I can leave some comments there

File: .github/workflows/build.yml (hunk 4 above)
Comment (levieggertcru) [reply .github/workflows/build.yml:31]: Hey @frett no preferred reason here, I hadn't thought about using pod install, but will try that, one less lane to have.

File: .github/workflows/build.yml (hunk 2 above)
Comment (frett) [reply .github/workflows/build.yml:45]: we've done it for some Android projects, let me get you a link to a sample

Comment (frett) [reply .github/workflows/build.yml:45]: https://github.com/CruGlobal/godtools-android/blob/develop/.github/workflows/build.yml#L115-L118

Comment (frett) [reply .github/workflows/build.yml:45]: We can generate a new App Store API key from their UI so you don't have to worry about extracting the current one. let me know if you need help with this or can't access it.

File: .github/workflows/build.yml (hunk 4 above)
Comment (frett) [reply .github/workflows/build.yml:31]: Using more direct steps when fastlane isn't actually required (e.g. cocoapods) makes it much more understandable for users who don't know fastlane.

**Comment by frett** [comment]
@levieggertcru it would be beneficial to include this cancel previous jobs job: https://github.com/CruGlobal/godtools-android/blob/develop/.github/workflows/build.yml#L10-L17

that way if there are multiple new commits to a branch/pr it will avoid duplicated work by GHA

File: fastlane/.env.default [hunk 5]
```diff
@@ -1,25 +1,29 @@
...
-CRU_TEST_DEVICES="iPhone 7"
 
-CRU_DEFAULT_BRANCH="master"
 
-## AppStore release related variables
-CRU_API_KEY_PATH="./fastlane/AppleAppStoreAPI.json"
-CRU_IDFA_IS_ENABLED=true
-CRU_IDFA_TRACKS_INSTALL=true
+APPLE_DEVELOPER_TEAM_ID = "DQ48D9BF2V"
+APPLE_ID_USERNAME = "cruapps@cru.org"
```
Comment (frett) [inline fastlane/.env.default:4]: as far as I know, this option is no longer required now that we are using an API key

File: .github/workflows/build.yml (hunk 1 above)
Comment (levieggertcru) [reply .github/workflows/build.yml]: thanks @frett , I will remove this step

File: godtools/App/Services/Renderer/BackgroundImageRenderer/MobileContentBackgroundImageRendererType.swift [hunk 6]
```diff
@@ -6,7 +6,7 @@
 //  Copyright © 2021 Cru. All rights reserved.
 //
 
-import Foundation
+import UIKit
```
Comment (frett) [inline godtools/App/Services/Renderer/BackgroundImageRenderer/MobileContentBackgroundImageRendererType.swift:9]: it looks like there was some code updates that got included with this PR. I'm guessing that wasn't intentional.

Are these updates something that should be handled in a different PR, or are they changes that don't matter being bundled here?

Comment (frett) [reply godtools/App/Services/Renderer/BackgroundImageRenderer/MobileContentBackgroundImageRendererType.swift:9]: nm, I see the commit. you were fixing the tests

File: .github/workflows/build.yml (hunk 3 above)
Comment (levieggertcru) [reply .github/workflows/build.yml:52]: @frett just created a PR for those new lanes

File: .github/workflows/build.yml (hunk 2 above)
Comment (levieggertcru) [reply .github/workflows/build.yml:48]: @frett I think I can use this instead https://docs.fastlane.tools/actions/app_store_connect_api_key/

How would we go about generating a new App Store API key and using that?

Comment (frett) [reply .github/workflows/build.yml:48]: that is what the `AppleAppStoreApi.json` key is ;) it's found on app store connect under "Users and Access" -> "Keys"

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -40,12 +46,11 @@ jobs:
...
-      - name: Decrypt AppleAppStoreApi.json.enc For Fastlane App Store Connect Access 
+      - name: Import App Store Connect API Key
         if: github.event_name == 'push' && github.ref == 'refs/heads/master'
-        env: 
-          KEY: ${{ secrets.ENCRYPTED_CF3A694B4ADE_KEY }}
-          IV: ${{ secrets.ENCRYPTED_CF3A694B4ADE_IV }}
-        run: openssl aes-256-cbc -K ${KEY} -iv ${IV} -in fastlane/AppleAppStoreApi.json.enc -out fastlane/AppleAppStoreApi.json -d
+        env:
+          APP_STORE_CONNECT_API_JSON_PAYLOAD: ${{ secrets.APP_STORE_CONNECT_API_JSON_PAYLOAD }}
+        run: echo $APP_STORE_CONNECT_API_JSON_PAYLOAD > fastlane/AppleAppStoreApi.json
```
Comment (frett) [inline .github/workflows/build.yml:53]: 🎉

**Review by frett — APPROVED** [review]

**Review by frett — APPROVED** [review]
